client/flaschen.py

```diff
@@ -0,0 +1,64 @@
+# -*- mode: c++; c-basic-offset: 4; indent-tabs-mode: nil; -*-
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation version 2.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://gnu.org/licenses/gpl-2.0.txt>
+
+import socket
+
+class Flaschen(object):
+  '''A Framebuffer display interface that sends a frame via UDP.'''
+
+  def __init__(self, host, port, width, height, layer=0):
+    self.host = host
+    self.port = port
+    self.width = width
+    self.height = height
+    self.layer = layer
+    self.pixels = []
+    for x in xrange(width):
+      self.pixels.append([(0, 0, 0) for y in xrange(height)])
+    self.sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
+
+  def _header(self):
+    return ''.join(["P6\n",
+                    "45 35\n",
+                    "255\n"])
+
+  def _footer(self):
+    return ''.join(["0\n",
+                    "0\n",
+                    "%d\n" % self.layer])
+
+  def set(self, x, y, color, transparent=False):
+    '''Set the pixel at the given coordinates to the specified color.
+
+    Args:
+      x: x offset of the pixel to set
+      y: y offset of the piyel to set
+      color: A 3 tuple of (r, g, b) color values, 0-255
+      transparent: If true, black(0, 0, 0) will be transparent and show the layer below.
+    '''
+    if x >= self.width or y >= self.height:
+      return
+    if color == (0, 0, 0) and transparent:
+      color = (1, 1, 1)
+    self.pixels[x][y] = color
+  
+  def send(self):
+    '''Send the updated pixels to the display.'''
+    data = []
+    for y in xrange(0, 35):
+      for x in xrange(0, 45):
+        data.append(''.join([chr(c) for c in self.pixels[x][y]]))
```

Comment:
How efficient is this operation ? It sounds like a lot of memory allocation churn to append() to an unknown size (realloc() happening all the time underneath) unless Python does some smart buffer-reuse and stuff underneath.

It is probably a good idea to pre-allocate that buffer once per object (including prepended header and footer) and have pixels being set in-place on set() directly at the place where it happens (so essentally at data_ + header-offset.

 In particular for the video-display situation with a lot of refresh, this can probably cut down on the CPU-use a couple of factors.

Might or might not be problem - if you measured it and deemed ok, then this is probably fine. Your call.
At least - is there a way to hint Python what size the data will be (I have never programmed Python, so I don't know) ? data = [] sounds like it starts with a buffer of 0, while we would like to hint it hat the final data size will be size.width * size.height * 3